qiskit/qasm3/grammar.py

```diff
@@ -0,0 +1,644 @@
+
+
+class QuantumMeasurementAssignment(Statement):
+    """
+    quantumMeasurementAssignment
+        : quantumMeasurement ARROW indexIdentifierList
+        | indexIdentifier EQUALS quantumMeasurement  # eg: bits = measure qubits;
+    """
+
+    def __init__(self, indexIdentifier: IndexIdentifier2, quantumMeasurement: QuantumMeasurement):
+        self.indexIdentifier = indexIdentifier
+        self.quantumMeasurement = quantumMeasurement
+
+    def qasm(self):
+        return [f"{self.indexIdentifier.qasm()} = {self.quantumMeasurement.qasm()}"]
+
+
+class ExpressionTerminator(Expression):
+    """
+    expressionTerminator
+        : Constant
+        | Integer
+        | RealNumber
+        | booleanLiteral
+        | Identifier
+        | StringLiteral
+        | builtInCall
+        | kernelCall
+        | subroutineCall
+        | timingTerminator
+        | LPAREN expression RPAREN
+        | expressionTerminator LBRACKET expression RBRACKET
+        | expressionTerminator incrementor
+    """
+
+    pass
+
+
+class Integer(Expression):
+    """Integer : Digit+ ;"""
+
+    pass
+
+
+class Designator(Class):
+    """
+    designator
+        : LBRACKET expression RBRACKET
+    """
+
+    def __init__(self, expression: Expression):
+        self.expression = expression
+
+    def qasm(self):
+        return f"[{self.expression.qasm()}]"
+
+
+class BitDeclaration(Class):
+    """
+    bitDeclaration
+        : ( 'creg' Identifier designator? |   # NOT SUPPORTED
+            'bit' designator? Identifier ) equalsExpression?
+    """
+
+    def __init__(self, identifier: Identifier, designator=None, equalsExpression=None):
+        self.identifier = identifier
+        self.designator = designator
+        self.equalsExpression = equalsExpression
+
+    def qasm(self):
+        return [f"bit{self.designator.qasm()} {self.identifier.qasm()};\n"]
+
+
+class QuantumDeclaration(Class):
+    """
+    quantumDeclaration
+        : 'qreg' Identifier designator? |   # NOT SUPPORTED
+         'qubit' designator? Identifier
+    """
+
+    def __init__(self, identifier: Identifier, designator=None):
+        self.identifier = identifier
+        self.designator = designator
+
+    def qasm(self):
+        return [f"qubit{self.designator.qasm()} {self.identifier.qasm()};\n"]
+
+
+class QuantumGateCall(QuantumInstruction):
+    """
+    quantumGateCall
+        : quantumGateModifier* quantumGateName ( LPAREN expressionList? RPAREN )? indexIdentifierList
+    """
+
+    def __init__(
+        self,
+        quantumGateName: Identifier,
+        indexIdentifierList: [Identifier],
+        expressionList: [Expression] = None,
+        quantumGateModifier=None,
+    ):
+        self.quantumGateName = quantumGateName
+        self.indexIdentifierList = indexIdentifierList
+        self.expressionList = expressionList
+        self.quantumGateModifier = quantumGateModifier
+
+    def qasm(self):
```

Comment:
It looks like `quantumGateModifier`s are not yet supported. Should we warn/raise in that case?

Reply:
Solved in 1ucian0#8